cunumeric/deferred.py

```diff
@@ -441,27 +688,63 @@ def set_item(self, key, rhs):
         assert self.dtype == rhs.dtype
         # Check to see if this is advanced indexing or not
         if is_advanced_indexing(key):
+            view_copy = False
             # Create the indexing array
-            index_array = self._create_indexing_array(key)
-            if index_array.shape != rhs.shape:
-                raise ValueError(
-                    "Advanced indexing array does not match source shape"
-                )
-            if self.ndim != index_array.ndim:
-                raise NotImplementedError(
-                    "need support for indirect partitioning"
+            copy_needed, store, index_array = self._create_indexing_array(
+                key, True
+            )
+            if copy_needed:
+                if self.base.transform.bottom:
+                    lhs = self
+                else:
+                    # if  store is transformed we need to to return a copy of
+                    # the store since Copy operation can't be done on
+                    # the store with transformation
+                    store_to_copy = DeferredArray(
+                        self.runtime,
+                        base=store,
+                        dtype=self.dtype,
+                    )
+                    store_copy = self.runtime.create_empty_thunk(
+                        store_to_copy.shape,
+                        self.dtype,
+                        inputs=[store_to_copy],
+                    )
+                    store_copy.copy(store_to_copy, deep=True)
+
+                    lhs = store_copy
+                    view_copy = True
+            else:
+                lhs = self
+                view_copy = False
+
+            if rhs.ndim == 0:
+                rhs_tmp = self.runtime.create_empty_thunk(
+                    index_array.base.shape,
+                    self.dtype,
+                    inputs=[index_array],
```

Comment:
I don't think the contents of `rhs_tmp` here depend on `index_array`.

Reply:
right, it doesn't